CompilerSource/parsing/pretty_printer.cpp

```diff
@@ -0,0 +1,1343 @@
+  } else if (node.expression->type == AST::NodeType::FUNCTION_CALL) {
+    VisitFunctionCallExpression(*node.expression->As<FunctionCallExpression>());
+  } else if (node.expression->type == AST::NodeType::TERNARY_EXPRESSION) {
+    VisitTernaryExpression(*node.expression->As<TernaryExpression>());
+  } else if (node.expression->type == AST::NodeType::SIZEOF) {
+    VisitSizeofExpression(*node.expression->As<SizeofExpression>());
+  } else if (node.expression->type == AST::NodeType::ALIGNOF) {
+    VisitAlignofExpression(*node.expression->As<AlignofExpression>());
+  } else if (node.expression->type == AST::NodeType::CAST) {
+    VisitCastExpression(*node.expression->As<CastExpression>());
+  } else if (node.expression->type == AST::NodeType::ARRAY) {
+    VisitArray(*node.expression->As<Array>());
+  }
+}
+
+bool AST::Visitor::VisitFullType(FullType &ft, bool print_type) {
+  if (print_type) {
+    std::vector<std::size_t> flags_values = {jdi::builtin_flag__const->value,    jdi::builtin_flag__static->value,
+                                             jdi::builtin_flag__volatile->value, jdi::builtin_flag__mutable->value,
+                                             jdi::builtin_flag__register->value, jdi::builtin_flag__inline->value,
+                                             jdi::builtin_flag__Complex->value,  jdi::builtin_flag__unsigned->value,
+                                             jdi::builtin_flag__signed->value,   jdi::builtin_flag__short->value,
+                                             jdi::builtin_flag__long->value,     jdi::builtin_flag__long_long->value,
+                                             jdi::builtin_flag__restrict->value, jdi::builtin_typeflag__override->value,
+                                             jdi::builtin_typeflag__final->value};
+
+    std::vector<std::size_t> flags_masks = {
+        jdi::builtin_flag__const->mask,    jdi::builtin_flag__static->mask,       jdi::builtin_flag__volatile->mask,
+        jdi::builtin_flag__mutable->mask,  jdi::builtin_flag__register->mask,     jdi::builtin_flag__inline->mask,
+        jdi::builtin_flag__Complex->mask,  jdi::builtin_flag__unsigned->mask,     jdi::builtin_flag__signed->mask,
+        jdi::builtin_flag__short->mask,    jdi::builtin_flag__long->mask,         jdi::builtin_flag__long_long->mask,
+        jdi::builtin_flag__restrict->mask, jdi::builtin_typeflag__override->mask, jdi::builtin_typeflag__final->mask};
+
+    std::vector<std::string> flags_names = {"const",  "static",    "volatile", "mutable",  "register",
+                                            "inline", "complex",   "unsigned", "signed",   "short",
+                                            "long",   "long long", "restrict", "override", "final"};
+
+    for (std::size_t i = 0; i < flags_values.size(); i++) {
+      if ((ft.flags & flags_masks[i]) == flags_values[i]) {
+        if (i != 8 || (i == 8 && ft.def->name == "char")) {
```

Comment:
In most cases, `signed` is redundant, but it is most commonly used with `char`.

Reply:
Heh, I don't think we need to include engineers' paranoia in this.  However, if you want to, please use the constant name rather than hard-coding `8` in the source.

The general rule of thumb is that no numbers should appear in the code outside of a constant definition except 0 and 1 (and even then, if there's a constant that more gracefully captures why you're using that number, you should prefer it...).

Reply:
The purpose of this check is for testing. Imagine that I would need to add `signed` everywhere I used a type.